### PR TITLE
core: go runtime stats

### PR DESCRIPTION
**Pull request by mcutalo88**
### Description

Go runtime stat reporter. This can be enabled with existence of `go_runtime_stats`, it's useful to be able to disable this during development as it can be quite noisy.

### Testing Performed

Unit / locally

```
gateway:
...
  stats:
    flush_interval: 1s
    log_reporter: {}
    go_runtime_stats:
      collection_interval: 1s
```

```
gateway:
...
  stats:
    flush_interval: 1s
    log_reporter: {}

    // Defaults to 10 second collection interval
    go_runtime_stats:{}
```
```
2020-11-10T10:44:26.270-0800	DEBUG	stats/debug_reporter.go:41	counter	{"name": "clutch.gateway.start", "value": 1, "tags": {}, "type": "counter"}
2020-11-10T10:44:26.270-0800	DEBUG	stats/debug_reporter.go:50	gauge	{"name": "clutch.runtime.memory.gc.count", "value": 5, "tags": {}, "type": "gauge"}
2020-11-10T10:44:26.270-0800	DEBUG	stats/debug_reporter.go:50	gauge	{"name": "clutch.runtime.cpu.goroutines", "value": 4, "tags": {}, "type": "gauge"}
2020-11-10T10:44:26.270-0800	DEBUG	stats/debug_reporter.go:50	gauge	{"name": "clutch.runtime.cpu.cgo_calls", "value": 1, "tags": {}, "type": "gauge"}
2020-11-10T10:44:26.270-0800	DEBUG	stats/debug_reporter.go:50	gauge	{"name": "clutch.runtime.memory.malloc", "value": 171739, "tags": {}, "type": "gauge"}
2020-11-10T10:44:26.270-0800	DEBUG	stats/debug_reporter.go:50	gauge	{"name": "clutch.runtime.memory.heap.idle", "value": 52576256, "tags": {}, "type": "gauge"}
2020-11-10T10:44:26.270-0800	DEBUG	stats/debug_reporter.go:50	gauge	{"name": "clutch.runtime.memory.othersys", "value": 1974365, "tags": {}, "type": "gauge"}
2020-11-10T10:44:26.270-0800	DEBUG	stats/debug_reporter.go:50	gauge	{"name": "clutch.runtime.memory.gc.next", "value": 16798496, "tags": {}, "type": "gauge"}
2020-11-10T10:44:26.270-0800	DEBUG	stats/debug_reporter.go:50	gauge	{"name": "clutch.runtime.memory.gc.last", "value": 1605033865300644000, "tags": {}, "type": "gauge"}
2020-11-10T10:44:26.270-0800	DEBUG	stats/debug_reporter.go:50	gauge	{"name": "clutch.runtime.memory.gc.pause_total", "value": 826256, "tags": {}, "type": "gauge"}
2020-11-10T10:44:26.270-0800	DEBUG	stats/debug_reporter.go:50	gauge	{"name": "clutch.runtime.memory.alloc", "value": 10397992, "tags": {}, "type": "gauge"}
2020-11-10T10:44:26.270-0800	DEBUG	stats/debug_reporter.go:50	gauge	{"name": "clutch.runtime.memory.total", "value": 22648816, "tags": {}, "type": "gauge"}
2020-11-10T10:44:26.271-0800	DEBUG	stats/debug_reporter.go:50	gauge	{"name": "clutch.runtime.memory.frees", "value": 120740, "tags": {}, "type": "gauge"}
2020-11-10T10:44:26.271-0800	DEBUG	stats/debug_reporter.go:50	gauge	{"name": "clutch.runtime.memory.heap.released", "value": 49569792, "tags": {}, "type": "gauge"}
2020-11-10T10:44:26.271-0800	DEBUG	stats/debug_reporter.go:50	gauge	{"name": "clutch.runtime.memory.stack.sys", "value": 688128, "tags": {}, "type": "gauge"}
2020-11-10T10:44:26.271-0800	DEBUG	stats/debug_reporter.go:50	gauge	{"name": "clutch.runtime.memory.stack.mcache_sys", "value": 16384, "tags": {}, "type": "gauge"}
2020-11-10T10:44:26.271-0800	DEBUG	stats/debug_reporter.go:50	gauge	{"name": "clutch.runtime.memory.stack.mspan_inuse", "value": 272544, "tags": {}, "type": "gauge"}
2020-11-10T10:44:26.271-0800	DEBUG	stats/debug_reporter.go:50	gauge	{"name": "clutch.runtime.memory.sys", "value": 76366848, "tags": {}, "type": "gauge"}
2020-11-10T10:44:26.271-0800	DEBUG	stats/debug_reporter.go:50	gauge	{"name": "clutch.runtime.memory.lookups", "value": 0, "tags": {}, "type": "gauge"}
2020-11-10T10:44:26.271-0800	DEBUG	stats/debug_reporter.go:50	gauge	{"name": "clutch.runtime.memory.heap.alloc", "value": 10397992, "tags": {}, "type": "gauge"}
2020-11-10T10:44:26.271-0800	DEBUG	stats/debug_reporter.go:50	gauge	{"name": "clutch.runtime.memory.heap.sys", "value": 66355200, "tags": {}, "type": "gauge"}
2020-11-10T10:44:26.271-0800	DEBUG	stats/debug_reporter.go:50	gauge	{"name": "clutch.runtime.memory.heap.inuse", "value": 13778944, "tags": {}, "type": "gauge"}
2020-11-10T10:44:26.271-0800	DEBUG	stats/debug_reporter.go:50	gauge	{"name": "clutch.runtime.memory.heap.objects", "value": 50999, "tags": {}, "type": "gauge"}
2020-11-10T10:44:26.271-0800	DEBUG	stats/debug_reporter.go:50	gauge	{"name": "clutch.runtime.memory.stack.inuse", "value": 688128, "tags": {}, "type": "gauge"}
2020-11-10T10:44:26.271-0800	DEBUG	stats/debug_reporter.go:50	gauge	{"name": "clutch.runtime.memory.stack.mspan_sys", "value": 311296, "tags": {}, "type": "gauge"}
2020-11-10T10:44:26.271-0800	DEBUG	stats/debug_reporter.go:50	gauge	{"name": "clutch.runtime.memory.stack.mcache_inuse", "value": 13888, "tags": {}, "type": "gauge"}
2020-11-10T10:44:26.271-0800	DEBUG	stats/debug_reporter.go:50	gauge	{"name": "clutch.runtime.memory.gc.sys", "value": 5564880, "tags": {}, "type": "gauge"}
2020-11-10T10:44:36.271-0800	DEBUG	stats/debug_reporter.go:50	gauge	{"name": "clutch.runtime.memory.othersys", "value": 1893461, "tags": {}, "type": "gauge"}
2020-11-10T10:44:36.271-0800	DEBUG	stats/debug_reporter.go:50	gauge	{"name": "clutch.runtime.memory.gc.next", "value": 16798496, "tags": {}, "type": "gauge"}
2020-11-10T10:44:36.271-0800	DEBUG	stats/debug_reporter.go:50	gauge	{"name": "clutch.runtime.memory.gc.count", "value": 5, "tags": {}, "type": "gauge"}
```

### CODE REVIEW

File: backend/gateway/stats/runtime.go [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+	r.scope.Gauge("cpu.goroutines").Update(float64(runtime.NumGoroutine()))
+	r.scope.Gauge("cpu.cgo_calls").Update(float64(runtime.NumCgoCall()))
+}
+
+func (r *RuntimeStatCollector) collectMemStats() {
+	var memStats runtime.MemStats
+	runtime.ReadMemStats(&memStats)
+
+	// General Stats
+	r.scope.Gauge("memory.alloc").Update(float64(memStats.Alloc))
```
Comment (danielhochman) [inline backend/gateway/stats/runtime.go:46]: you should predefine all of the gauges in a struct e.g.

```go
type runtimeMetrics struct {
   memoryAlloc tally.Gauge
}
```

and then set it up in New()

File: backend/gateway/gateway.go [hunk 2]
```diff
@@ -170,6 +170,12 @@ func RunWithConfig(f *Flags, cfg *gatewayv1.Config, cf *ComponentFactory, assets
 	}
 	ctx := context.TODO()
 
+	// Start collecting go runtime stats if enabled
+	if cfg.Gateway.Stats != nil && cfg.Gateway.Stats.CollectGoRuntimeStats {
+		runtimeStats := stats.NewRuntimeStats(scope)
+		go runtimeStats.Collect(ctx)
+	}
+
```
Comment (danielhochman) [inline backend/gateway/gateway.go:178]: id move this after as much of the set up as possible

**Review by danielhochman — COMMENTED** [review]
good overall, minor changes

File: backend/gateway/stats/runtime.go [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+		memoryGCCount:      runtimeScope.Gauge("memory.gc.count"),
+	}
+
+	return &RuntimeStatCollector{
+		runtimeMetrics: runtimeMetrics,
+	}
+}
+
+func (r *RuntimeStatCollector) Collect(ctx context.Context) {
+	ticker := time.NewTicker(time.Second * 10)
```
Comment (danielhochman) [inline backend/gateway/stats/runtime.go:91]: should this interval be configurable?

Comment (mcutalo88) [reply backend/gateway/stats/runtime.go:91]: Sure I have no objections to that, ill rework the config a bit.

**Review by danielhochman — APPROVED** [review]

**Pull request merged**